esp/bindings/http/platform/httpservice.cpp

```diff
@@ -974,3 +875,558 @@ int CEspHttpServer::onGet()
+        if (!isEmptyString(_userName))
+            userName.set(_userName);
+        else
+            userName.set((authReq.requestParams) ? authReq.requestParams->queryProp("username") : nullptr);
+        if (!isEmptyString(_password))
+            password.set(_password);
+        else
+            password.set((authReq.requestParams) ? authReq.requestParams->queryProp("password") : nullptr);
+        if (userName.isEmpty() || password.isEmpty())
+        {
+            ESPLOG(LogMin, "Authentication failed: invalid credential for session:<%u>", sessionID);
+            handleAuthFailed(true, authReq, &sessionID);
+            commitAndCloseRemoteConnection(conn);
+            return authFailed;
+        }
+
+        authReq.ctx->setUserID(userName.str());
+        authReq.ctx->setPassword(password.str());
+        authReq.authBinding->populateRequest(m_request.get());
+        if (!authReq.authBinding->doAuth(authReq.ctx))
+        {
+            ESPLOG(LogMin, "Authentication failed for session:<%u>", sessionID);
+            handleAuthFailed(true, authReq, &sessionID);
+            commitAndCloseRemoteConnection(conn);
+            return authFailed;
+        }
+
+        // authenticate optional groups
+        authOptionalGroups(authReq);
+
+        sessionTree->setProp(PropSessionUserID, userName.str());
+        sessionTree->addPropInt(PropSessionState, HTTPSS_active);
+        authReq.ctx->addUserToken(sessionID);
+    }
+    else
+    {
+        authOptionalGroups(authReq);
+
+        //The UserID has to be set before the populateRequest() because the UserID is used to create the user object.
+        //After the user object is created, we may call addUserToken().
+        userName.set(sessionTree->queryProp(PropSessionUserID));
+        authReq.ctx->setUserID(userName.str());
+        authReq.authBinding->populateRequest(m_request.get());
+        authReq.ctx->addUserToken(sessionID);
+    }
+
+    ESPLOG(LogMax, "Authenticated for %s@%s", userName.str(), sessionTree->queryProp(PropSessionNetworkAddress));
+    postSessionAuth(authReq, sessionID, sessionState, timeNow, conn, sessionTree);
+
+    return authSucceeded;
+}
+
+void CEspHttpServer::postSessionAuth(EspAuthRequest& authReq, unsigned sessionID, HTTPSessionState sessionState, time_t timeNow,
+    void* _conn, IPropertyTree* sessionTree)
+{
+    if (!_conn)
+        throw MakeStringException(-1, "Invalid SDS connection.");
+
+    IRemoteConnection* conn = (IRemoteConnection*) _conn;
+    if (authReq.methodName && strieq(authReq.methodName, "logout"))
+    {//delete this session before logout
+        IPropertyTree* root = conn->queryRoot();
+        if (!root)
+            throw MakeStringException(-1, "Failed to get SDS DomainSession.");
+
+        VStringBuffer path("%s[%s='%d']", PathSessionSession, PropSessionID, sessionID);
+        Owned<IPropertyTreeIterator> it = root->getElements(path.str());
+        ForEach(*it)
+            root->removeTree(&it->query());
```

Comment:
I think this is potentially problematic.. deleting elements while iterating like this.
Normal practice is to add to an array, then delete.